p4-16/spec/P4-16-spec.mdk

```diff
@@ -5228,6 +5228,36 @@ types (e.g. `structs`) are copied recursively, and all components
 of `header`s are copied, including "validity" bits. Assignment is
 not defined for `extern` values.
 
+### Compound assignment { #sec-compound-assignment }
+
+Compound-assignment operators provide a shorter syntax for assigning
+the result of an arithmetic or bitwise operator. The compound assignment
+operator expressions have the form `LHS op RHS`, where `op` is one of
+`+=`, `-=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `<<=`, `>>=`, and `LHS`,
+`RHS` are expressions with arithmetic types. The following table shows
+all P4 assignment operators and their equivalent.
+
+|-------------|----------------------------------|-----------|---------------|
+| Operator    | Operator name                    | Example   | Equivalent of |
++-------------+:--------------------------------:+:---------:+:-------------:+
+| `=`         | basic assignment                 | `a = b`   | NA            |
+| `+=`        | addition assignment              | `a += b`  | `a = a + b`   |
+| `-=`        | subtraction assignment           | `a -= b`  | `a = a - b`   |
+| `*=`        | multiplication assignment        | `a *= b`  | `a = a * b`   |
+| `/=`        | division assignment              | `a /= b`  | `a = a / b`   |
+| `%=`        | modulo assignment                | `a %= b`  | `a = a % b`   |
+| `&=`        | bitwise AND assignment           | `a &= b`  | `a = a & b`   |
+| `|=`        | bitwise OR assignment            | `a |= b`  | `a = a | b`   |
+| `^=`        | bitwise XOR assignment           | `a ^= b`  | `a = a ^ b`   |
+| `<<=`       | bitwise left shift assignment    | `a <<= b` | `a = a << b`  |
+| `>>=`       | bitwise right shift assignment   | `a >>= b` | `a = a >> b`  |
+|-------------|----------------------------------|-----------|---------------|
+
+The behavior of every builtin compound-assignment expression with the form `E1 op= E2`,
+where `E1` is a modifiable lvalue expression and `E2` is an rvalue expression,
+is exactly the same as the behavior of the expression `E1 = E1 op E2`.
```

Comment:
there's the point of side-effects in E1. Do they happen once or twice?
You could also specify that the operations are defined for all types where the expanded form is defined.

Reply:
I believe someone else mentioned this in another issue or PR, but explicitly saying that expression E1 is evaluated exactly once seems far preferable to saying that it is evaluated twice, or leaving the number of evaluations unspecified.

Reply:
I have added the following paragraph (similar to the [notes for assignment in C](https://en.cppreference.com/w/c/language/operator_assignment#Notes)) to describe side-effects in `E1`:

> The side effect of updating `E1` is sequenced after the value computations, but not the side effects of `E1` and `E2` themselves and the evaluations of the operands are, as usual, unsequenced relative to each other.